planner/core/rule_eliminate_projection.go

```diff
@@ -160,17 +161,27 @@ func (pe *projectionEliminator) eliminate(p LogicalPlan, replace map[string]*exp
 	return p.Children()[0]
 }
 
+func init() {
+	// This makes us able to use ReplaceColumnOfExpr in expression package and avoid dependency cycle.
+	expression.ReplaceColumnOfExpr = ReplaceColumnOfExpr
+}
+
 // ReplaceColumnOfExpr replaces column of expression by another LogicalProjection.
-func ReplaceColumnOfExpr(expr expression.Expression, proj *LogicalProjection, schema *expression.Schema) expression.Expression {
+func ReplaceColumnOfExpr(expr expression.Expression, projExprs []expression.Expression, schema *expression.Schema) expression.Expression {
 	switch v := expr.(type) {
 	case *expression.Column:
 		idx := schema.ColumnIndex(v)
-		if idx != -1 && idx < len(proj.Exprs) {
-			return proj.Exprs[idx]
+		if idx != -1 && idx < len(projExprs) {
+			return projExprs[idx]
 		}
 	case *expression.ScalarFunction:
 		for i := range v.GetArgs() {
-			v.GetArgs()[i] = ReplaceColumnOfExpr(v.GetArgs()[i], proj, schema)
+			v.GetArgs()[i] = ReplaceColumnOfExpr(v.GetArgs()[i], projExprs, schema)
+		}
+		// We have done special optimization for IN expression in PR#13594. This optimization makes us need to do the replacement
+		// for some more expressions other than builtinFunc.args
+		if v.FuncName.L == ast.In {
```

Comment:
Can we override the `getArgs()` function for `baseInSig` to return the `nonConstArgs`?

Reply:
Seems we can't again.
https://github.com/pingcap/tidb/blob/3cf4214167a360fa8fd14b119a10fce68ffe9bb8/expression/builtin_other.go#L237-L241
`baseBuiltinFunc.args` is still being used so we can't return only `nonConstArgs`.
If we return `baseBuiltinFunc.args + nonConstArgs`. That will cause `getArgs()` to allocate a new slice to hold them, then we can't use the result of `getArgs()` to modify the original slices.

Reply:
Can we introduce another interface function `setArgs` for `builtinFunc` and specially implement it for `baseInSig`? What I am concerned is that, there may be something like projection elimination in other places, or in future code changes, then this function would be ignored once again when processing its arguments, so we'd better encapsulate the arguments layer, instead of specially handle them every time.